src/triage/util/db.py

```diff
@@ -34,3 +36,17 @@ def __reconstruct__(cls, url, creator, kwargs):
 
 
 create_engine = SerializableDbEngine
+
+
+@contextmanager
+def scoped_session(db_engine):
```

Comment:
I think the existing uses of sessionmaker in Triage are probably mistaken.  We aren't really binding anything to it except for the engine. I copied the sessionmaker usage from some example somewhere I'm sure, thinking that was *the* way to make sessions (I wasn't coming into this with a ton of prior sqlalchemy knowledge). But if all we are binding to the sessionmaker is an engine, we can just skip that and call Session's constructor with the engine directly when we need it. If we're using sessionmaker just for the sake of it (and I know a lot of the code is doing this at present), I just think we're adding an unnecessary layer that makes the code harder to follow.

Applying this to your example, I think we would basically skip the sessionmaker property, make the scoped_session method take an engine, and implement the sessionmanager/get_for_update the way you have there. What do you think of this?

Reply:
yeah that makes sense to me.

if we expected to need further tweaks to session creation, (and certainly if we didn't ever want the raw engine / connection), then it might still make sense to _pass around_ a session maker.

but from the sound of it, we really want the DB engine to be the parameter, at least much of the time, and it's easy enough to instantiate a session from it, without quite so much complexity/overhead as we have now.

and certainly if we consolidated some of this as much as possible -- that is, made it DRY, with such useful methods on Experiment, or whatever -- then yes I think that probably would help as well.